File: Codechef-2019/JUN19B/SUMAGCD.cpp
#include <bits/stdc++.h>
using namespace std;

long long maxi=-1, gcd, sz, a[100009], lf[100009], rt[100009], n, t, x;

int main()  {
    cin>>t;
    set<long long> s;
    while(t--)  {
        s.clear();
        maxi=-1;
        cin>>n;
        int j=1;

        for(int i=0;i<n;i++){
            cin>>x;
            sz=s.size();
            s.insert(x);
            if(sz!=s.size())    {
                a[j]=x;
                j++;
            }
        }
        j--;
        for(int k=0; k<=j; k++){
          cout<<a[k]<<"  ";
        }
        cout<<endl;

        lf[0]=-1;
        rt[j+1]=-1;

        lf[1]=a[1];
        rt[j]=a[j];
        for(int k=0; k<=j; k++){
          cout<<rt[k]<<"  ";
        }
        cout<<endl;

        for(int k=0; k<=j; k++){
          cout<<lf[k]<<"  ";
        }
        cout<<endl;

        for(int i=2;i<=j;i++)
            lf[i]=__gcd(lf[i-1],a[i]);
        for(int i=j-1;i>=1;i--)
            rt[i]=__gcd(rt[i+1],a[i]);
        for(int i=2;i<j;i++)    {
            gcd=__gcd(lf[i-1],rt[i+1]) + a[i];
            maxi=max(maxi,gcd);
        }
        for(int k=0; k<=j; k++){
          cout<<rt[k]<<"  ";
        }
        cout<<endl;

        for(int k=0; k<=j; k++){
          cout<<lf[k]<<"  ";
        }
        cout<<endl;

        if(lf[0]==-1 && rt[2]==-1)
            maxi=max(maxi,2*a[0]);
        else    {
            maxi=max(maxi,a[1]+rt[2]);
            maxi=max(maxi,a[j]+lf[j-1]);
        }
        if(j==1)
            maxi=max(maxi,2*a[1]);
        cout<<maxi<<endl;
    }
    return 0;
}

File: Codechef-2019/JUN19B/SUMAGCD.py
# -*- coding: utf-8 -*-
"""
@author: anishukla
"""

""" Link to the problem: 
    https://www.codechef.com/JUNE19B/problems/SUMAGCD
    Link to my profile:
    https://www.codechef.com/users/an1shukla """

import numpy as np

def gcd(a, b):
    if b == 0:
        return a
    else:
        return gcd(b, a % b)

T = int(input())
for i in range(T):
    N = int(input())
    A = list(map(int, input().split()))
    
    A = np.unique(A)
    l = len(A)
    A = list(A)
    A.insert(0, 0)
    req = -1
    
    start = [0]*(l+2)
    end = [0]*(l+2)
    
    start[0] = -1
    start[1] = A[1]
    end[l+1] = -1
    end[l] = A[l]
    
    for i in range(2, l+1):
        start[i] =  gcd(start[i-1], A[i])
        
    for i in range(l-1, 0, -1):
        end[i] =  gcd(end[i+1], A[i])
        
    for i in range(2, l):
        res =  gcd(start[i-1], end[i+1]) +A[i]
        
        req = max(res, req)
        
    if(start[0] == -1 and end[2] == -1):
        req = max(req, 2*A[0])
    
    else:
        req = max(req, A[1]+end[2])
        req = max(req, A[l] + start[l-1])
        
    if(l == 1):
        req = max(req, 2*A[1])
    
    print(req)
